src/GovUk.Education.ExploreEducationStatistics.Admin/Security/AuthorizationHandlers/AuthorizationHandlerService.cs

```diff
@@ -123,4 +135,58 @@ public class AuthorizationHandlerResourceRoleService
 
         return usersReleaseRoles.Any(releaseRoles.Contains);
     }
+
+    public async Task<bool> IsReleaseViewableByUser(Release release, ClaimsPrincipal user)
+    {
+        // If the user has the "AccessAllReleases" Claim, they can see any Release.
+        if (SecurityUtils.HasClaim(user, SecurityClaimTypes.AccessAllReleases))
+        {
+            return true;
+        }
+
+        // If the user has any PublicationRoles on the owning Publication, they can see its child Releases.
+        if (await HasRolesOnPublication(
+                    user.GetUserId(),
+                    release.PublicationId,
+                    EnumUtil.GetEnumValuesAsArray<PublicationRole>()))
+        {
+            return true;
+        }
+
+        // If the user has any non-Pre-release Viewer roles on the Release, they can see it at any time.
+        if (await HasRolesOnRelease(
+                    user.GetUserId(),
+                    release.Id,
+                    UnrestrictedReleaseViewerRoles))
+        {
+            return true;
+        }
+
+        // If the user has the Pre-release Viewer role on this Release and the Release is within its open
+        // Pre-release window, they can see the Release.
+        if (await HasRolesOnRelease(
+                    user.GetUserId(),
+                    release.Id,
+                    ReleaseRole.PrereleaseViewer))
+        {
+            var windowStatus = _preReleaseService.GetPreReleaseWindowStatus(release, DateTime.UtcNow);
+            if (windowStatus.Access == PreReleaseAccess.Within)
+            {
+                return true;
+            }
+        }
+
+        await _contentDbContext
+            .Entry(release)
+            .Reference(r => r.Publication)
+            .LoadAsync();
+
+        await _contentDbContext
+            .Entry(release.Publication)
+            .Collection(p => p.Releases)
+            .LoadAsync();
+
+        // If the Release is public, anyone can see it.
+        return release.IsLatestPublishedVersionOfRelease();
+    }
```

Comment:
Do you think we should add some unit tests for this so that other handlers can just mock this service and rely on it already being tested?

Reply:
Yeah this was added at a time whereby we were pushing for more integration in our unit tests, so actually there wasn't any solid need to create an interface.  I think that the only definite reasons we need to add interfaces is for mocking and replacing with alternative implementations for certain use cases.

Ironically, I remember thinking at the time that this was a very good instance where I thought that mocks would be better than integration as it would make the handler tests much more concise and the expectations would actually confer a lot of meaning in these cases, but we were moving in the other direction at the time.

This class when used in unit tests does accept mocks within itself though sometimes so that we don't have to add too much standing data for release and publication roles though.

Happp to fall in line with convention and add an interface if you think it needs one, but right now there's no *actual* need for one as far as I know.

Reply:
Ok yep, happy to leave it as it is. I see the usages of this are well tested themselves so no need to switch up the tests as they currently stand.

Reply:
Cool ta